### DISCUSSION TITLE
How to parse overlapping keywords?

### DISCUSSION THREAD

**Discussion by msrd0** [zesterer/chumsky#580] [Q&A]
I'm trying to parse a file that contains keywords that might overlap, i.e. `whilelse` is valid and needs to parsed the same as `while else`. Ideally I wouldn't hardcode those valid combinations but infer them programatically. I tried looking for a parser that supports peeking but couldn't find anything.

<details><summary>Basically I'm trying to use chumsky to replace this manual parser:</summary>

_AoC Spoiler Warning_
https://msrd0.dev/msrd0/aoc23/src/commit/9cf3b3ef76c762491fd4fcb980522918be0f074c/src/bin/day1.rs#L77-L98

</details>

**Comment by zesterer** [comment]
You can do peeking with `and_is` or `rewind`. That said, the wider problem sounds rather complex to solve and somewhat outside the scope of chumsky, so you might want to use `custom` for this case.
